Docs 4426 kryo v4.2 (#961)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/configure-custom-serializers.adoc

= Configure Custom Serializers
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> You can change the serialization mechanism to improve functionality and performance when Mule runtime engine executes any of the following processes:

Reviewer: change > customize (?)

> By default, Mule runtime engine uses plain old Java serialization. You can configure `defaultObjectSerializer` in your Mule application to specify a different serialization mechanism such as the Kryo serializer or any other custom serializer.

Reviewer 2: ```suggestion
By default, Mule runtime engine (Mule) uses ordinary Java serialization. However, you can configure `defaultObjectSerializer` in your Mule application to specify a different serialization mechanism, such as the Kryo serializer or any other custom serializer.
```
...but I personally really loved "plain old"...!  :-)

Reviewer 2: The previous text mentions configuring defaultObjectSerializer as the way to specify Kyro or API.
But then we have Configuring topics for Kyro and API...followed by "Obtaining" the defaultObjectSerializer.
I am confused about the user task order; do or should the two Configure topics include configuring/obtaining the default value?

* Read from or write to a persistent ObjectStore
* Read from or write to a persistent VM or JMS queue

> * An object distributes through a Mule cluster

Reviewer: An object distributes through a Mule cluster > Distributes an object through a Mule cluster (Only if Mule process does this) If not leave it as it is.

* Read from or write to an object from a file

== Compatibility

If you are running Mule runtime engine (Enterprise Edition), you can configure the Kryo serializer to improve performance.

> If you are using Mule Kernel(Community Edition), you can create a custom serializer using the Serialization API.

Reviewer 3: ```suggestion
If you are using Mule Kernel (Community Edition), you can create a custom serializer using the Serialization API.
```

> The Batch module always uses the Kryo serializer, so the module's behavior is not affected by the configured serializer.

Reviewer 2: ```suggestion
Customization does not affect the behavior of the Batch module, which always uses the Kryo serializer.
```

> == Compatibility

Reviewer 2: I wonder if we should start making titles like this one more specific to the actual content, in preparation for potential reuse? We must have hundreds of "Compatibility" titles, so we would consider something like Compatibility Between Serializers and Mule.

If you are running Mule runtime engine (Enterprise Edition), you can configure the Kryo serializer to improve performance.

If you are using Mule Kernel (Community Edition), you can create a custom serializer using the Serialization API.

> 

Reviewer 2: OR...maybe the previous title should be "Configure the Serializer Compatible with Your Implementation of Mule"...because the previous topic actually introduces (contains) the two following task topics:

==Compatibility (Kyro or API)
===Kyro
===API

== Configure the Kryo Serializer

> Mule provides an implementation of `ObjectSerializer` which relies on the Kryo framework. Using Kryo provides the following benefits:

Reviewer 2: ```suggestion
Mule provides an implementation of `ObjectSerializer` that relies on the Kryo framework. Using Kryo provides the following benefits:
```

* Better performance +

> Kryo is much faster than Java serialization

Reviewer: Add . (dot) at the end of sentence. _Kryo is much faster than Java serialization._

> * Support for a wider range on Java types +

Reviewer 2: ```suggestion
* Support for a wider range of Java types +
```

> Kryo is not bounded by most of the limitations that Java serialization imposes like requiring to implement the `Serializable` interface, having a default constructor, etc.

Reviewer 2: ```suggestion
Kryo is not bound by most of the limitations that Java serialization imposes, such as requiring the implementation of the `Serializable` interface, having a default constructor, and so on.
```

* Support for compression +
You can use either Deflate or GZip compression algorithms.

> The Kryo namespace enables you to configure this serializer inside your Mule application, without defining a custom Spring bean. The following configuration example sets the default serializer to a Kryo based one:

Reviewer 2: ```suggestion
The Kryo namespace enables you to configure this serializer inside your Mule application, without defining a custom Spring bean. The following configuration example sets the default serializer to a Kryo-based one:
```

[source,xml,linenums]
----
<?xml version="1.0" encoding="UTF-8"?>
<mule xmlns="http://www.mulesoft.org/schema/mule/core"
       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xmlns:kryo="http://www.mulesoft.org/schema/mule/kryo"
       xsi:schemaLocation="
        http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
        http://www.mulesoft.org/schema/mule/kryo http://www.mulesoft.org/schema/mule/kryo/current/mule-kryo.xsd">

    <kryo:serializer name="kryo" />
    <configuration defaultObjectSerializer-ref="kryo" />

</mule>
----

> Additionally, you can also include the `compressionMode` XML attribute to configure compression:

Reviewer 2: We don't need both "additionally" and "also":
```suggestion
You can also include the `compressionMode` XML attribute to configure compression:
```

[source,xml,linenums]
----
<kryo:serializer name="noCompression" compressionMode="NONE" /> <!-- NONE is the default value -->
<kryo:serializer name="deflate" compressionMode="DEFLATE" />
<kryo:serializer name="gzip" compressionMode="GZIP" />
----

> == Serialization API

Reviewer 2: I suggest omitting this as a separate topic and move it to be the introduction to Configure a Custom Serializer Using the Serialization API.

> The open source Serialization API provides the `ObjectSerializer.java` interface which enables you to serialize or deserialize objects into a byte array. See https://github.com/mulesoft/mule-api/blob/master/src/main/java/org/mule/runtime/api/serialization/ObjectSerializer.java[ObjectSerializer.java in GitHub].

Reviewer 2: Dang it! Grammarly wants us to hyphenate open source.... IMHO, it doesn't much matter, except if there might be confusion, but we should prolly follow our tool.... Apologies. At least my explanation was worthwhile....
```suggestion
The open-source Serialization API provides the `ObjectSerializer.java` interface, which enables you to serialize or deserialize objects into a byte array. See https://github.com/mulesoft/mule-api/blob/master/src/main/java/org/mule/runtime/api/serialization/ObjectSerializer.java[ObjectSerializer.java in GitHub].
```

Reviewer 2: Move to follow the next title....

The Serialization API has the following features:

* Is thread-safe
* Passes an OutputStream when serializing and streaming
* Allows an InputStream as an input source

If you are going to serialize and deserialize in Mule, then you must define the `getInternalProtocol()` method.
If you are going to use the serialized object in an external system, define that in the `getExternalProtocol()` method. +

> A https://github.com/mulesoft/mule-api/blob/master/src/main/java/org/mule/runtime/api/serialization/SerializationProtocol.java[Serialization Protocol] is the Object used to serialize and deserialize.

Reviewer 2: ```suggestion
A https://github.com/mulesoft/mule-api/blob/master/src/main/java/org/mule/runtime/api/serialization/SerializationProtocol.java[Serialization Protocol] is the object used to serialize and deserialize.
```

> A https://github.com/mulesoft/mule-api/blob/master/src/main/java/org/mule/runtime/api/serialization/SerializationProtocol.java[Serialization Protocol] is the Object used to serialize and deserialize.

Reviewer 2: move

> == Configure a Custom Serializer

Reviewer 2: Oh! Now I see the peer title to Configure the Kyro Serializer! So...I suggest moving the text from the previous topic here instead.

> == Configure a Custom Serializer

Reviewer 2: ```suggestion
== Configure a Custom Serializer Using the Serializer API
```

> Configure the default `ObjectSerializer` for your Mule application using the Mule `<configuration>` tag:

Reviewer: Does it works here to also mention the property `defaultObjectSerializer` that is inside the tag? This property was also mentioned in the first section of the doc.  _You can configure defaultObjectSerializer in your Mule application to specify a different serialization mechanism such as the Kryo serializer or any other custom serializer._ Because you have the `ObjectSerializer` property and also the `defaultObjectSerializer` one, and was not sure if were different ones.

<mule xmlns="http://www.mulesoft.org/schema/mule/core"
       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xmlns:spring="http://www.springframework.org/schema/beans"
       xsi:schemaLocation="
        http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
        http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-current.xsd">

  <spring:config name="Spring_Config" files="serializer.xml" />
  <configuration defaultObjectSerializer-ref="customSerializer" />

</mule>
----

Then, configure the bean in the `serializer.xml` file:

[source,xml,linenums]
----
<?xml version="1.0" encoding="utf-8"?>
<beans xmlns="http://www.springframework.org/schema/beans"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-4.3.xsd">
	<bean id="customSerializer" name="customSerializer"

> 		class="com.my.CustomSerializer" scope="singleton">

Reviewer 3: ```suggestion
		class="com.my.CustomSerializer">
```
(my bad, copy-pasting from an example)

	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-4.3.xsd">
	<bean id="customSerializer" name="customSerializer"
		class="com.my.CustomSerializer">
		<!-- ... -->
	</bean>
</beans>
----

> == Obtain a Configured ObjectSerializer

Reviewer 2: I'm confused about the placement of this topic. 
Why is it placed following the Kyro and API config topics; isn't it PART of those topics?
What is the chronology/user journey?

> There are different ways to obtain an `ObjectSerializer` inside Java code. Generally, the recommended approach is through dependency injection. The following code example shows you how to get the currently configured `ObjectSerializer`:

Reviewer 2: Why does the user have to obtain this; isn't it part of the previous two tasks?
Or does this belong before those two tasks?

[source,java,linenums]
----
public class MyClass {
   @Inject
   private  ObjectSerializer objectSerializer;
}
----

If you want a specifically named serializer (whether it's the default or not), you can access the serializer by name:

[source,java,linenums]
----
public class MyClass {
  @Inject
  @Named ("mySerializer")
  private ObjectSerializer objectSerializer;
}
----

> == Performance Improvements with Kryo

Reviewer 2: Isn't this partially redundant of both the topics set intro text and the text in the Kyro config topic?
Try to deduplicate the topic set.
Perhaps this belongs before Configure Kyro, or maybe as part of a Kyro set of topics?

> Using Kryo provides particular performance improvements over plain old Java serialization when you are using these components:

Reviewer 2: ```suggestion
Using Kryo provides particular performance improvements over ordinary Java serialization when you are using these components:
```

> * Persistent  or clustered object stores

Reviewer 2: Remove extra space
```suggestion
* Persistent or clustered object stores
```

* Persistent or distributed VM queues
* JMS connector

> === Compression

Reviewer 2: Consider avoiding this subtitle.  If the primary advantage of using Kyro is that it does not use compression, then maybe the title of this topic should include reference to compression.
And maybe the topic set intro should mention it as well.

> The Kryo serializer without compression is significantly faster than the plain old Java serializer in all cases. However, the compression mode only provides an actual improvement on the high availability (HA) cases.

Reviewer 2: Is the Kyro serializer always without compression, or are there two versions, with and without?
Again, I think we might want to mention "no compression" as nearly synonymous with Kyro....

Reviewer 2: ```suggestion
The Kryo serializer without compression is significantly faster than the ordinary Java serializer in all cases. However, compression mode provides an improvement only in high availability (HA) cases.
```

> For the compression to be worthy, the amount of time the CPU spends compressing and decompressing has to be significantly lower than the amount of I/O time saved by reducing the payload size. Because network operations are typically slower than disk operations and because HA clustering requires node replication, which translates to more traffic, only in the HA case the compression paid off.

Reviewer 2: ```suggestion
For compression to be worthwhile, the amount of time the CPU spends compressing and decompressing has to be significantly lower than the amount of I/O time saved by reducing the payload size. Because network operations are typically slower than disk operations, and because HA clustering requires node replication (which translates to more traffic), compression is useful only in HA environments.
```

> This is not a universal constant. You might be running Mule on machines with slower disks or higher I/O demands in which compression might be worthy in any case. Also, these tests were performed with 1 MB payloads, but the larger the data stream, the more worthy becomes the compression.

Reviewer 2: ```suggestion
This is not a universal constant. You might be running Mule on machines with slower disks or higher I/O demands in which compression might be worthwhile in any case. Also, tests were performed using 1 MB payloads. If your data stream is larger, compression becomes more worthwhile.
```

> == Limitations and Considerations

Reviewer 2: Requirements and Restrictions?

> Ensure that you are aware of the following limitations and considerations when using serializers.

Reviewer 2: ```suggestion
Before you commit to using custom serialization, ensure that you are aware of the following limitations and considerations when using serializers.
```
I wonder if this information belongs at the beginning, before actually performing the configurations (know before you do)?

I wonder if this whole topic set is really "Compare Default Java Serialization with Customized Serialization," maybe with subtopics that compare Java with Kyro in its specific Mule edition, and Java with API in its edition?

Just something to think about; I know it would be a big reorg....

=== Changing Serializers Requires a Clean Slate

> Serializers are not interoperable nor interchangeable. That means that if you decide to change the serializer your application uses, you need to make sure that all messages in VM and JMS queues have been consumed and that those queues are empty by the time the new serializer kicks in. This is because Kryo serializer won’t be able to read datagrams written by the Java serializer and vice-versa. The same thing applies to persistent ObjectStores. If you try to read an entry generated with a different serializer, you can't read the entry.

Reviewer 2: In the following text, are we referring to generic object stores or the ObjectStore database?
```suggestion
Serializers are not interoperable nor interchangeable. That means that if you decide to change the serializer your application uses, you need to make sure that all messages in VM and JMS queues have been consumed and that those queues are empty by the time the new serializer takes over. The Kryo serializer won’t be able to read datagrams written by the Java serializer, and vice-versa. The same thing applies to persistent object stores; you cannot read with one serializer an entry generated in a different serializer.
```

=== Serialization in a Shared VM Connector

> Domains provide a way to share resources between applications. For example, you can define a VM connector on a domain to allow inter-app communication through VM message queues. However, serializers can only be configured at an application level, they cannot be configured at a domain. If applications A and B communicate with each other through a VM connector defined on a domain to which both belong, but A serializes using Java and B using Kryo, it just works. Whenever either app tries to write to an endpoint that uses the shared connector, that particular message is not serialized with the application’s serializer but the one the VM connector uses. This is good for the plug and play experience, but you won’t be able to tell that shared VM connector to use Kryo and get a performance improvement out of it.

Reviewer 2: I suggest that this long paragraph be divided into two or more.
I had trouble following the example, so I am not sure how to edit it.
I don't understand "it just works"; does that mean that this works only under specific circumstances?
The following edit is copyediting only; I don't have the tech experience to suggest rewording. Perhaps ask a peer to look at this?

```suggestion
Domains provide a way to share resources between applications. For example, you can define a VM connector on a domain to allow inter-app communication through VM message queues. 

However, serializers can be configured only at an application level; they cannot be configured at a domain. 

If applications A and B communicate with each other through a VM connector defined on a domain to which both belong, but A serializes using Java and B using Kryo, it just works. Whenever either app tries to write to an endpoint that uses the shared connector, that particular message is not serialized with the application’s serializer but the one the VM connector uses. This is good for a "plug-and-play experience," but you won’t be able to realize a performance improvement by having that shared VM connector use Kryo.
```

> === Less Improvement for a Local Persistent ObjectStore

Reviewer 2: ObjectStore commercial object database, or a generic object store? GLOBAL

Reviewer 2: ```suggestion
=== Less Improvement Using a Local Persistent ObjectStore
```

> Unlike the other cases, the local persistent ObjectStore doesn’t show much improvement because of high contention on the ObjectStore implementation which pretty much absorbs all the gain.

Reviewer 2: ```suggestion
Using the local persistent ObjectStore has no performance advantage because of high contention on the ObjectStore implementation.
```
Do we really need a whole topic for this one point?
Maybe these subtopics should be bullet-list items? Hmm...most are pretty long, however....

=== No JMS Improvement Chart

> Per the JMS API, the queues don’t work with raw payload objects. Instead, you have to provide an instance of the `javax.jms.Message` class. The broker client is then responsible for serializing it, not Mule. Therefore, the impact of Kryo in such a scenario is minimum. The only performance gain of using Kryo with JMS is that Mule serializes the MuleSession and puts it as a header in Base64 format. Serializing the MuleSession with Kryo can give you up to 10% performance speed, but we don’t consider it as an example use case since the big part of the serialization is up to the JMS broker instead of Mule.

Reviewer 2: What does this mean, "per the JMS API"? Do we mean "When using the JMS API"?
"the queues" = which queues?

Reviewer 2: Can we reduce this topic as follows and still retain value to the audience?
```suggestion
The only performance gain of using Kryo with JMS is that Mule serializes the MuleSession instance and uses it as a header in Base64 format. Serializing MuleSession with Kryo can give you up to a 10% performance increase, but this is primarily attributable to the JMS broker instead of Mule.
```

> Per the JMS API, the queues don’t work with raw payload objects. Instead, you have to provide an instance of the javax.jms.Message class. The broker client is then responsible for serializing it, not Mule. Therefore, the impact of Kryo in such a scenario is minimum. The only performance gain of using Kryo with JMS is that Mule serializes the MuleSession and puts it as a header in Base64 format. Serializing the MuleSession with Kryo can give you up to 10% performance speed, but we don’t consider it as an example use case since the big part of the serialization is up to the JMS broker instead of Mule.

Reviewer: `javax.jms.Message`

=== Problematic Types

> Although Kryo is capable of serializing objects that don’t implement the `Serializable` interface, setting Kryo as the default serializer doesn’t mean that components such as the VM connector, ObjectSerializer, or Cluster can handle objects which don’t implement such an interface. That’s because even though Kryo can deal with those objects, the Java APIs for those components still expect instances of Serializable in their method signatures. +

Reviewer 2: Copyedit only
```suggestion
Although Kryo is capable of serializing objects that don’t implement the `Serializable` interface, setting Kryo as the default serializer doesn’t mean that components such as the VM connector, ObjectSerializer, or cluster can handle objects that don’t implement such an interface. Although Kryo can work with those objects, the Java APIs for those components still expect instances of `Serializable` in their method signatures. +
```

> Although Kryo is capable is serializing objects that don’t implement the Serializable interface, setting Kryo as the default serializer doesn’t mean that components such as the VM connector, ObjectSerializer, or Cluster can handle objects which don’t implement such an interface. That’s because even though Kryo can deal with those objects, the Java APIs for those components still expect instances of Serializable in their method signatures. +

Reviewer: capable _of_ serializing

> Plain old Java serialization fails with an object that implements the Serializable interface. However, if serialization contains another object which doesn’t implement the Serializable interface, Kryo is likely (but not guaranteed) to succeed. A typical case is a POJO containing an `org.apache.xerces.jaxp.datatype.XMLGregorianCalendarImpl,` which is in use in the https://www.anypoint.mulesoft.com/exchange/?search=netsuite[NetSuite] or https://www.anypoint.mulesoft.com/exchange/?search=Microsoft+Dynamics+CRM&type=connector[Microsoft Dynamics CRM] connectors.

Reviewer 2: ```suggestion
Ordinary Java serialization fails with an object that implements the Serializable interface. However, if serialization contains another object that doesn’t implement the Serializable interface, Kryo is likely (but not guaranteed) to succeed. A typical case is a POJO containing an `org.apache.xerces.jaxp.datatype.XMLGregorianCalendarImpl,` which is in use in the https://www.anypoint.mulesoft.com/exchange/?search=netsuite[NetSuite] or https://www.anypoint.mulesoft.com/exchange/?search=Microsoft+Dynamics+CRM&type=connector[Microsoft Dynamics CRM] connectors.
```
